### PR TITLE
feat: Accept nullable values in "tags" HashMap in `Add` action

### PR DESCRIPTION
**Pull request by gotocoding-DB**
## What changes are proposed in this pull request?

Align the schema of AddAction with the schema in Scala implementation of Delta: https://github.com/delta-io/delta/blob/11873fe40f6fe7660b63b2c7611df40330865620/kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/AddFile.java#L59-L61

To achieve this, we need to support `Option<HashMap<K, V>>` in `allow_null_container_values` macros.

## How was this change tested?

We had a test where we check the expected types. I changed the expected value for tags' values from "required" to "nullable".

### CODE REVIEW

**Comment by gotocoding-DB** [comment]
Please take a look at this PR: @nicklan, @zachschuermann, and @scovich (I don't have the rights to add you as reviewers).

File: derive-macros/src/lib.rs [hunk 1]
```diff
@@ -119,9 +119,35 @@ fn gen_schema_fields(data: &Data) -> TokenStream {
                     }
                 });
                 if have_schema_null {
-                    if let Some(last_ident) = type_path.path.segments.last().map(|seg| &seg.ident) {
-                        if last_ident != "HashMap" {
-                           return Error::new(
+                    // Check if the type is HashMap or Option<HashMap>
```
Comment (gotocoding-DB) [inline derive-macros/src/lib.rs]: An alternative is to simply remove this check entirely (the part with `if let Some(last_ident) = type_path.path.segments.last().map(|seg| &seg.ident)`).

File: derive-macros/src/lib.rs [hunk 2]
```diff
@@ -119,9 +119,35 @@ fn gen_schema_fields(data: &Data) -> TokenStream {
...
+                                })
+                            } else {
+                                false
+                            }
+                        } else {
+                            false
+                        }
+                    } else {
+                        false
+                    };
```
Comment (scovich) [inline derive-macros/src/lib.rs]: Mutable values can help simplify the logic a lot:
```suggestion
                    let mut is_valid = false; // supported cases will change it to true below
                    if let Some(seg) = type_path.path.segments.last() {
                        if seg.ident == "HashMap" {
                            is_valid = true;
                        } else if seg.ident == "Option" {
                            // Check if Option wraps a HashMap
                            if let PathArguments::AngleBracketed(args) = &seg.arguments {
                                is_valid = matches!(
                                    args.first(), // Option accepts exactly one template arg
                                    Some(syn::GenericArgument::Type(Type::Path(inner)))
                                    if inner.path.segments.last().is_some_and(|s| s.ident == "HashMap")
                                )
                            }
                        }
                    }
```

File: derive-macros/src/lib.rs [hunk 3]
```diff
@@ -119,9 +119,35 @@ fn gen_schema_fields(data: &Data) -> TokenStream {
...
+
+                    if !is_valid {
+                        if let Some(last_ident) = last_segment.map(|seg| &seg.ident) {
```
Comment (scovich) [inline derive-macros/src/lib.rs]: Why would it be ok to swallow the error if `last_segment` is None?
Seems like that's still an error case and we just need to construct a more robust error message?

File: derive-macros/src/lib.rs (hunk 2 above)
Comment (scovich) [reply derive-macros/src/lib.rs]: ... but it's probably better to factor this out as a helper function, because the indentation is becoming extreme:
```rust
if !allows_null_container_values(type_path) {
    return Error::new(...);
}
```

where
```rust
fn allows_null_container_values(type_path: &syn::TypePath) -> bool {
    if let Some(seg) = type_path.path.segments.last() {
        if seg.ident == "HashMap" {
            return true;
        } else if seg.ident == "Option" {
            // Check if Option wraps a HashMap
            if let PathArguments::AngleBracketed(args) = &seg.arguments {
                return matches!(
                    args.first(), // Option accepts exactly one template arg
                    Some(syn::GenericArgument::Type(Type::Path(inner)))
                    if inner.path.segments.last().is_some_and(|s| s.ident == "HashMap")
                )
            }
        }
    }
    false
}
```

**Comment by scovich** [comment]
It seems like we need some unit tests for this change? Negative tests can be tricky (maybe best avoided), but we should at least have positive tests that the newly allowed types and schemas are actually allowed and behaving as expected?

File: derive-macros/src/lib.rs (hunk 2 above)
Comment (gotocoding-DB) [reply derive-macros/src/lib.rs]: Thanks for the comment, now it looks much better!
Moved it to a separate function and tried to polish it, making it easier for the reader to understand what's happening inside.

File: derive-macros/src/lib.rs (hunk 3 above)
Comment (gotocoding-DB) [reply derive-macros/src/lib.rs]: We shouldn't allow None of course.
Now we return an error: `"Can only use allow_null_container_values on Option<HashMap> fields"`

**Comment by gotocoding-DB** [comment]
@scovich I uploaded a completely new version that doesn't modify derive-macros at all because I realized the initial approach was incorrect. We need to read NULL values and keep them in "tags," but with `#[allow_null_container_values]`, we lose them after parsing the JSON. Please review the test I've added; I tried to be very detailed and include some kind of contract there.

**Comment by gotocoding-DB** [comment]
@scovich So few more findings after your comment.

> but we will also need to update the row visitors because they don't support nullable map keys today

That's right, but we don't read tags at all: https://github.com/delta-io/delta-kernel-rs/blob/main/kernel/src/actions/visitors.rs#L120.
So, changing the schema of AddAction shouldn't break anything in `RowVisitor`?

Second thing. I also realized that other actions also have "tags":
* `Remove`: https://github.com/delta-io/delta-kernel-rs/blob/main/kernel/src/actions/mod.rs#L786. This one is the same as Add and also nullable in Delta-java: https://github.com/delta-io/delta/blob/master/kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/RemoveFile.java#L40 
* [CheckpointMetadata](https://github.com/delta-io/delta-kernel-rs/blob/main/kernel/src/actions/mod.rs#L921), [Cdc](https://github.com/delta-io/delta-kernel-rs/blob/main/kernel/src/actions/mod.rs#L835), [Sidecar](https://github.com/delta-io/delta-kernel-rs/blob/main/kernel/src/actions/mod.rs#L884)

File: kernel/src/actions/mod.rs [hunk 4]
```diff
@@ -1974,4 +1974,51 @@ mod tests {
         assert!(record_batch.column(2).is_null(0));
         assert!(record_batch.column(3).is_null(0));
     }
+
+    #[test]
+    fn test_add_tags_deserialization() {
```
Comment (DrakeLin) [inline kernel/src/actions/mod.rs]: nit: I would separate these cases into 3 unit tests.

Comment (gotocoding-DB) [reply kernel/src/actions/mod.rs]: Sure, I'll do it

**Comment by gotocoding-DB** [comment]
@scovich Hey Ryan, can you please take a look at my response above, please: https://github.com/delta-io/delta-kernel-rs/pull/1395#issuecomment-3406241673?

If I'm missing something, or if changing the tags' schema doesn't make sense without supporting them in RowVisitor, I can also try adding parsing tags into RowVisitor. I just don't want to merge everything into a single PR if it can be two separate parts, but if you think it's better to combine both parts into one PR, I will do that!

File: kernel/src/actions/mod.rs (hunk 4 above)
Comment (gotocoding-DB) [reply kernel/src/actions/mod.rs]: Fixed

**Comment by gotocoding-DB** [comment]
@DrakeLin Hey Drake, could you please take a look again? I addressed your comment.
